File: code/java/25.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverse(ListNode head){
        ListNode p = head;
        ListNode q = head.next;
        if(q==null)
            return head;
        head = q;
        p.next = null;
        while(head!=null){
            head = head.next;
            q.next = p;
            p = q;
            q = head;
        }
        return p;
    }

    public ListNode reverseKGroup(ListNode head, int k) {
        ListNode result = null;
        int count = 1;
        ListNode start = head;
        ListNode end = null;
        ListNode last = null;
        while(head!=null){
            if(count==k){
                end = head;
                head = head.next;
                end.next = null;
                end = reverse(start);
                if(result==null)
                    result = end;
                if(last==null){
                    last = start;
                }
                else{
                    last.next = end;
                    last = start;
                }
                start = head;
                count = 1;
            }
            else{
                count++;
                head = head.next;
            }
        }
        if(count>1){
            if(last==null)
                return start;
            last.next = start;
        }
        return result;
    }
}

File: code/python/25.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        def reverse(root, k):
            cur = root
            idx = 1
            while cur and idx < k:
                cur = cur.next
                idx += 1
            # 长度不够，直接返回
            if idx < k or cur is None:
                return root,root, None
            # 截断最后一个node与后面节点的连接
            post = cur.next
            cur.next = None
            # 头插法逆转当前截断的链表
            vituralNode = ListNode(-1)
            cur = root
            while cur:
                temp = cur.next
                cur.next = None
                cur.next = vituralNode.next
                vituralNode.next = cur
                cur = temp
            head = vituralNode.next
            vituralNode.next = None
            return head,root, post
        if not head or k == 0: return head
        res = []
        # 分段反转链表，长度不够就表示到头了
        while True:
            headN,endN,postN = reverse(head, k)
            res.append((headN,endN))
            if postN is None: break
            head = postN
        # 分段的链表头尾相连
        lastEnd = None
        for headN,endN in res:
            if lastEnd:
                lastEnd.next = headN
            lastEnd = endN
        return res[0][0]

